.gitignore

```diff
@@ -10,6 +10,9 @@ generated/
 training_results/
 train-output/
 
+# We use this directory for storing the data from functional tests.
+__functional-test-data/
```

Comment:
Tests should not write to the local directory. Please assume that the local directory is read-only for tests and that they can only write to a dedicated area.

Reply:
Where should we write instead?

Reply:
Maybe `/tmp`?

Reply:
Fixed.

Reply:
In shell scripts, use mktemp and then add a trap that removes the directory on exit.

Reply:
I'll do that instead, thanks!

Reply:
I disagree on the expectation to do a build in a read-only directory. A common practice for builds is to store output relative to the source (although using an ENV for redirection). Then you can inspect and access the content, and it's easy to know what (and for a build target to know what to clean). CI systems tend to place your working directory on a location that is ephermal and has fast storage, so when you start using non-relative you may be picking less optimal locations and could be making things more complicated.  As an example of the latter in CI build chains you may have a task that slurps up test output for formatting, but if its temp now you need logic to write out the location and find it. An example of the former is that tempfs is often undersized so if you have some significant test output it run out of storage. That said, CI scripts should reassign TMP/TMPDIR anyway.

Reply:
Actually we both agree on the same thing. I now see how my original statement was not clear enough and could be easily misunderstood. What I meant to say is that tests should not modify the source directory / local git clone. Tests should not add, remove, or modify files like code, tests, and test data.

Instead tests should isolate themselves from the project data and any other test case. This makes tests reproducible and allows parallel execution of tests. Our pytest-based test cases use fixtures for tmp path and file system isolation for a unique CWD for each test. pytest has [options](https://docs.pytest.org/en/stable/how-to/tmp_path.html#temporary-directory-location-and-retention) to retain temporary files, move its base directory to a different disk, and deal with distributed tests.

Your remarks about tempfs and TMP sound plausible, too. I rarely have to deal with these low level things and usually let the test runner take care of my environment.

Reply:
ah yes sorry for misunderstanding, you are completely right here thanks!

Reply:
I removed this directory from the .gitignore, and now both the functional test & e2e tests are using the mktemp + trap approach.